docs/architecture/psa-migration/md-cipher-dispatch.md

```diff
@@ -477,3 +503,57 @@ The architecture can be extended to support `MBEDTLS_PSA_CRYPTO_CLIENT` with a l
 
 * Compile-time dependencies: instead of checking `defined(MBEDTLS_PSA_CRYPTO_C)`, check `defined(MBEDTLS_PSA_CRYPTO_C) || defined(MBEDTLS_PSA_CRYPTO_CLIENT)`.
 * Implementers of `MBEDTLS_PSA_CRYPTO_CLIENT` will need to provide `psa_can_do_hash()` (or a more general function `psa_can_do`) alongside `psa_crypto_init()`. Note that at this point, it will become a public interface, hence we won't be able to change it at a whim.
+
+### Cipher light
+
+#### Definition
+
+**Note:** this definition is tentative an may be refined when implementing and
+testing, based and what's needed by internal users of Cipher light. The new
+config symbol will not be considered public so its definition may change.
+
+Cipher light will be automatically enabled in `build_info.h` by modules that
+need it, namely: CTR\_DRBG, CCM, GCM. Note: CCM and GCM currently depend on
+the full `CIPHER_C` (enforced by `check_config.h`); this hard dependency would
+be replaced by the above auto-enablement.
+
+Cipher light includes:
+- info functions;
+- support for block ciphers in ECB mode, encrypt only (note: in Cipher, "ECB"
+  means just one block, contrary to PSA);
+- the one-shot API as well as (part of) the streaming API;
+- only AES, Aria and Camellia.
+
+This excludes:
+- the AEAD/KW API (both one-shot and streaming);
+- support for stream ciphers;
+- support for other modes of block ciphers (CBC, CTR, CFB, etc.);
+- DES and variants (3DES).
+
+The following API functions, and supporting types, are candidates for
+inclusion in the Cipher light API, with limited features as above:
```

Comment:
Good question. I've dumped my thoughts on that in the latest commit, please have a look and let me know what you think. I have to say I'm not so sure. Right now my feeling is it depends if we really stop at GCM/CCM, or end up having time for more before 3.6 (which is obviously hard to predict).

Reply:
@gilles-peskine-arm did you have any chance to look at this? This looks like the most important unresolved question so far.

Reply:
For that sort of thing I'd want to prototype, I can't really tell by just staring at the code. I haven't done any prototyping and am unlikely to have the time soon.

Reply:
I'd be ok with just deciding that since we're unsure, we'll keep `mbedtls_cipher_info_t*` around because it's the predictable option: it might have a small cost, but it's low-risk.

Reply:
> For that sort of thing I'd want to prototype, I can't really tell by just staring at the code.

Good point.

> I haven't done any prototyping and am unlikely to have the time soon.

Well, I think that would be a bit beyond your duties as a reviewer anyway. But Valerio or I could do a prototype - or actually two I think.

Reply:
Some prototyping is going on at #8454 and #8466

Reply:
I think the prototypes have stabilized. The two approaches explored were:
- #8466: new internal module (different API & independent implementation);
- #8454: subset of cipher (API & implementation).

Using a new internal module has the following advantages:
- smaller code size: about 800 bytes;
- smaller RAM usage: about 60 bytes per GCM/CCM context (also, avoids a dynamic allocation);
- [perhaps subjective] the code is easier to modify than the existing `cipher.c` - this may matter when we add support for PSA dispatch (G5).

I think the only advantage I can see for the "subset" approach is that [_if_ we end up caring about CMAC](https://github.com/Mbed-TLS/mbedtls/pull/8057/files#r1381555015) (for people who don't have a CMAC driver, but only an AES-primitive driver or no driver at all), then only the "subset" approach will support code size reduction there, while the "new module" approach won't, because the CMAC API is too entangled with the Cipher API to be implemented on top of something else (CMAC functions take `mbedtls_cipher_context` and `mbedtls_cipher_info` pointers as arguments).